File: coauthor/prompt_utils.py
import os
import yaml
from jinja2 import Template
from typing import Dict, Any

from .file_utils import read_file
from .logging_utils import logger


def load_yaml(file_path: str) -> dict:
    """Load a YAML file and return its contents as a dictionary."""
    with open(file_path, "r") as f:
        return yaml.safe_load(f)


def merge_dicts(base: dict, override: dict) -> dict:
    """Merge two dictionaries recursively."""
    result = base.copy()
    for key, value in override.items():
        if isinstance(value, dict) and key in result:
            result[key] = merge_dicts(result[key], value)
        else:
            result[key] = value
    return result


def load_agent_settings_and_prompts(agent_path: str, agent: str):
    """Load agent settings and prompts from YAML files."""

    def load_agent_from_yaml(agent_path, agent_name):
        agent_file = f"{agent_path}/{agent_name}.yaml"
        if not os.path.exists(agent_file):
            raise FileNotFoundError(f"Task prompt file not found: {agent_file}")

        config = load_yaml(agent_file)
        parent = config.get("inherits")

        if parent:
            parent_settings, parent_prompts = load_agent_from_yaml(agent_path, parent)

            # Extract settings and prompts from current agent
            agent_settings = config.get("settings", {}) or {}
            agent_prompts = config.get("prompts", {}) or {}

            # Merge with parent settings and prompts
            settings = merge_dicts(parent_settings, agent_settings)
            prompts = merge_dicts(parent_prompts, agent_prompts)
        else:
            settings = config.get("settings", {}) or {}
            prompts = config.get("prompts", {}) or {}

            # Handle prefills if present in settings
            if "prefills" not in settings:
                settings["prefills"] = []

        return settings, prompts

    return load_agent_from_yaml(agent_path, agent)


def render_prompt(prompt: str, variables: Dict[str, Any]) -> str:
    """Render a prompt string using Jinja2 templating."""
    template = Template(prompt)
    rendered_prompt = template.render(**variables)
    # logger.info(f"Rendered prompt: {rendered_prompt}")
    return rendered_prompt


def get_xml_format_from_file(file: str) -> str:
    return f'<document name="{file}">\n' f"{read_file(file)}\n" f"</document>"


def get_xml_format_from_files(files: list[str]) -> str:
    return "\n".join(get_xml_format_from_file(file) for file in files) if files else None


def get_list_of_files(files: list[str] | None) -> str | None:
    if not files:  # Handles None and empty list cases
        return None

    # Filter out None values and convert all items to strings
    valid_files = [str(f) for f in files if f is not None]

    if not valid_files:  # Return None if no valid files after filtering
        return None

    return ", ".join(valid_files)

File: coauthor-for-vs-code/src/utils/promptUtils.ts
import * as nunjucks from 'nunjucks';
import { debug, error, initializeLogging } from '../logger/logUtils';
import { readFile } from './fileUtils';
import * as yaml from 'yaml';
import * as vscode from 'vscode';
import * as path from 'path';
import { getConfig } from './commonUtils';

const CHANNEL = 'PromptUtils';
initializeLogging(CHANNEL);

/**
 * Render a prompt string using nunjucks templating
 * @param prompt The prompt template string
 * @param variables Variables to use in template rendering
 * @returns Rendered prompt string
 */
export async function renderPrompt(
  prompt: string,
  variables: { [key: string]: any },
): Promise<string> {
  try {
    const env = nunjucks.configure({ autoescape: false });
    const renderedPrompt = env.renderString(prompt, variables);
    // debug(CHANNEL, `Rendered prompt: ${renderedPrompt}`);
    return renderedPrompt;
  } catch (err) {
    error(
      CHANNEL,
      `Error rendering prompt: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Get XML formatted string from a single file
 * @param file Path to the file
 * @returns XML formatted string containing file content
 */
export async function getXmlFormatFromFile(file: string): Promise<string> {
  try {
    const content = await readFile(file);
    return `<document name="${file}">\n${content}\n</document>`;
  } catch (err) {
    error(
      CHANNEL,
      `Error formatting file as XML: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Get XML formatted string from multiple files
 * @param files List of file paths
 * @returns XML formatted string containing all file contents, or null if no files
 */
export async function getXmlFormatFromFiles(
  files: string[],
): Promise<string | null> {
  try {
    if (!files || files.length === 0) {
      return null;
    }

    const xmlPromises = files.map((file) => getXmlFormatFromFile(file));
    const xmlContents = await Promise.all(xmlPromises);
    return xmlContents.join('\n');
  } catch (err) {
    error(
      CHANNEL,
      `Error formatting files as XML: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Convert a list of files to a comma-separated string
 * @param files List of file paths
 * @returns Comma-separated string of file paths
 */
export function getListOfFiles(files: string[] | null | undefined): string {
  try {
    if (!files) {
      return '';
    }
    return files.filter((f) => f !== null).join(', ');
  } catch (err) {
    error(
      CHANNEL,
      `Error creating file list: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Load a YAML file and return its contents as a dictionary
 * Handles both absolute paths and paths relative to extension's global storage
 * @param filePath Path to the YAML file
 * @param context Extension context (required for relative paths)
 * @returns Promise<object> Parsed YAML content
 */
export async function loadYaml(
  filePath: string,
  context?: vscode.ExtensionContext,
): Promise<object> {
  try {
    let absolutePath: string;
    const rootPath = getConfig<string>('explorer.rootPath', 'agents');

    if (path.isAbsolute(filePath)) {
      // If filePath is absolute, use it directly
      absolutePath = filePath;
    } else {
      // For any non-absolute path, use global storage as base
      if (!context) {
        throw new Error('Extension context required for relative paths');
      }

      try {
        const globalStoragePath = context.globalStorageUri.fsPath;
        const fullPath = path.join(globalStoragePath, rootPath, filePath);

        // Ensure the directory exists
        await vscode.workspace.fs.createDirectory(
          vscode.Uri.file(path.dirname(fullPath)),
        );
        debug(CHANNEL, `Using global storage path: ${fullPath}`);

        absolutePath = fullPath;
      } catch (err) {
        error(CHANNEL, `Error with global storage path: ${err}`);
        throw err;
      }
    }

    try {
      // Verify the path exists before trying to read it
      const fileUri = vscode.Uri.file(absolutePath);
      await vscode.workspace.fs.stat(fileUri);
      debug(CHANNEL, `Reading from: ${absolutePath}`);

      // Read and parse YAML
      const fileContent = await vscode.workspace.fs.readFile(fileUri);
      const yamlContent = Buffer.from(fileContent).toString('utf-8');
      const parsedYaml = yaml.parse(yamlContent);

      debug(CHANNEL, `Successfully loaded YAML from: ${filePath}`);
      return parsedYaml;
    } catch (err) {
      error(
        CHANNEL,
        `Path does not exist or is not accessible: ${absolutePath}`,
      );
      throw err;
    }
  } catch (err) {
    error(
      CHANNEL,
      `Error loading YAML file ${filePath}: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}

/**
 * Merge two dictionaries recursively
 * @param base Base dictionary
 * @param override Dictionary with overriding values
 * @returns Merged dictionary
 */
export function mergeDicts(
  base: { [key: string]: any },
  override: { [key: string]: any },
): { [key: string]: any } {
  try {
    const result = { ...base };
    for (const [key, value] of Object.entries(override)) {
      if (
        value &&
        typeof value === 'object' &&
        !Array.isArray(value) &&
        key in result
      ) {
        result[key] = mergeDicts(result[key], value);
      } else {
        result[key] = value;
      }
    }
    return result;
  } catch (err) {
    error(
      CHANNEL,
      `Error merging dictionaries: ${err instanceof Error ? err.message : String(err)}`,
    );
    throw err;
  }
}
